### PR TITLE
MM-41745 refactor upload license flow

### PR DESCRIPTION
**Pull request by pablovelezvidal**
#### Summary
Change the UI flow for the upload license process. This PR removes a step from the UI flow for uploading a license. With these changes when you click the upload button, the OS file selector window is shown and when you select the file THEN the upload modal is shown with the file information.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-41745

#### Screenshots

https://user-images.githubusercontent.com/10082627/156860041-010a3be3-3e4c-44e6-85d5-75d2d066e617.mp4

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: components/admin_console/license_settings/starter_edition/starter_left_panel.tsx [hunk 1]
```diff
@@ -73,14 +75,21 @@ const StarterLeftPanel: React.FC<StarterEditionProps> = ({
...
                     >
                         <FormattedMessage
                             id='admin.license.uploadFile'
                             defaultMessage='Upload File'
                         />
                     </button>
+                    <input
+                        ref={fileInputRef}
+                        type='file'
+                        accept='.mattermost-license'
```
Comment (julmondragon) [inline components/admin_console/license_settings/starter_edition/starter_left_panel.tsx:89]: This is also used in `upload_license_modal.tsx`, should be a constant config?

File: components/admin_console/license_settings/starter_edition/starter_left_panel.tsx [hunk 2]
```diff
@@ -73,14 +75,21 @@ const StarterLeftPanel: React.FC<StarterEditionProps> = ({
                 <div className='uploadButtons'>
                     <button
                         className='btn btn-upload light-blue-btn'
-                        onClick={openUploadModal}
+                        onClick={() => fileInputRef.current !== null && fileInputRef.current.click()}
```
Comment (julmondragon) [inline components/admin_console/license_settings/starter_edition/starter_left_panel.tsx:78]: What about `() => fileInputRef.current?.click()` ?

File: components/admin_console/license_settings/license_settings.tsx [hunk 3]
```diff
@@ -105,6 +106,20 @@ export default class LicenseSettings extends React.PureComponent<Props, State> {
...
+            this.props.actions.openModal({
+                modalId: ModalIdentifiers.UPLOAD_LICENSE,
+                dialogType: UploadLicenseModal,
+                dialogProps: {
+                    fileObjFromProps: this.state.file,
+                },
+            });
+        }
+        // eslint-disable-next-line react/no-did-update-set-state
+        this.setState({fileSelected: false, file: null});
```
Comment (julmondragon) [inline components/admin_console/license_settings/license_settings.tsx:120]: Does every props update need to clear the selected file state?

File: components/admin_console/license_settings/license_settings.tsx [hunk 4]
```diff
@@ -129,20 +144,22 @@ export default class LicenseSettings extends React.PureComponent<Props, State> {
         this.setState({upgradingPercentage: percentage || 0, upgradeError: error as string});
     }
 
+    handleChange = () => {
+        const element = this.fileInputRef.current;
+        if (element !== null && element.files !== null) {
+            if (element.files.length > 0) {
```
Comment (julmondragon) [inline components/admin_console/license_settings/license_settings.tsx:150]: What about `if (element?.files?.length) {` ?

**Review by julmondragon — APPROVED** [review]
Looks good, more straightforward flow.

**Comment by pablovelezvidal** [comment]
> Not sure if I understood that ticket right, but I thought the process would be:
> 
> * Click on the "Upload file" button from the page
> * User selects their license file
> * Show the user the "Successful upgrade" modal rather than the file selection modal.
> 
> Am I wrong?

hey @JulienTant the idea of the ticket is remove just one previous step where you got showed the selection file modal before to actually being able to select a file. So in the past the steps were:
- Click upload button
- modal select-file is shown
- click Choose file
- Select a file
-Get visual feedback about the file you want to upload

With this change the flow is:
- Click upload button
- Select a file
- modal select-file is shown with visual feedback about the file you want to upload

File: components/admin_console/license_settings/starter_edition/starter_left_panel.tsx (hunk 2 above)
Comment (pablovelezvidal) [reply components/admin_console/license_settings/starter_edition/starter_left_panel.tsx:78]: typescript linter was complained with that approach

File: components/admin_console/license_settings/starter_edition/starter_left_panel.tsx (hunk 1 above)
Comment (pablovelezvidal) [reply components/admin_console/license_settings/starter_edition/starter_left_panel.tsx:89]: it is a local constant in upload_license_modal, you are right, it is better if we set it as a global constant.

**Review by JulienTant — COMMENTED** [review]
Thanks for clarifying the workflow. It looks good to me @pablovelezvidal thanks!

**Review by JulienTant — APPROVED** [review]

**Review by stevemudie — APPROVED** [review]
Worked perfectly!

Looks good @pablovelezvidal!

**Review by annelieseklein — APPROVED** [review]
LGTM!

**Pull request merged**